feat: 공고 게시글 생성,수정,삭제 권한 로직 추가

sonypark opened the pull request:
## Resolve #85 

- 공고 게시글 생성,수정,삭제는 관리자만 할 수 있도록 해야한다.

## Changes

- `@AdminUser` 어노테이션을 통해 Notice 컨트롤러의 생성,수정,삭제 메서드는 관리자만 접근 가능하도록 수정했다.

## Notes

- Notice 컨트롤러에서  `@AdminUser`를 사용할 때 메서드 인자로 Boolean 타입을 받는데 사실상 필요없는 인자이다.  `@AdminUser`를 사용하기 위해 일다 남겨놨는데 이것이 최선일까..?
- 인수 테스트 코드에서 일반 사용자로 접근했을 때 예외를 발생시키는 테스트 코드도 추가해야한다.

## References

- N/A

sonypark committed:
feat: 토큰에 사용자 권한 정보 추가

sonypark committed:
feat: NoticeController에 AdminUser 어노테이션 추가

sonypark committed:
feat: 공고 조회 컨트롤러 메서드에 @NoValidate 추가

sonypark committed:
refactor: ResponseDTO lombok 생성자 접근 레벨 수정

sonypark committed:
feat: Role Enum 생성

c-yeonwoo commented on src/main/java/underdogs/devbie/auth/jwt/JwtTokenProvider.java:
```diff
@@ -30,8 +30,9 @@ public JwtTokenProvider(
     public String createToken(UserTokenDto userTokenDto) {
-        String subject = String.valueOf(userTokenDto.getId());
-        Claims claims = Jwts.claims().setSubject(subject);
+        Claims claims = Jwts.claims();
+        claims.put("userId", String.valueOf(userTokenDto.getId()));
+        claims.put("role", userTokenDto.getRole().toString());
```
오 이렇게하면 되는군요..ㅎㅎ

sonypark replied:
Thanks to @khb1109 👍

hongbin-dev replied:
✌

c-yeonwoo commented on src/main/java/underdogs/devbie/config/AuthConfig.java:
```diff
@@ -16,17 +17,18 @@
             .addPathPatterns("/api/**")
-            .excludePathPatterns("/api/auth/login", "/api/auth/login-url")
-            .excludePathPatterns("/api/notices", "/api/notices/**");
+            .excludePathPatterns("/api/auth/login", "/api/auth/login-url");
```
현재 인증이 필요한API와 그렇지 않은  API를 config에서 설정하는 방식과, @NoValidate를 사용하는 방식으로 혼재되어있는데, 팀원들과 이야기해서 하나로 통일하면 좋을듯하네요.

@NoValidate 어노테이션을 클래스에도 붙일 수 있으니, 코드를 볼때, 인증을 필요한지 아닌지를 어노테이션으로 바로 볼수 있다는점에서 후자가 더 편리할 것 같다는 생각이들긴합니다.

sonypark replied:
그렇네요. 동의합니다!
일단 로그인 컨트롤러 부분도 ` @novalidate를 사용하는 방식`으로 변경했습니다.

hongbin-dev commented on src/test/java/underdogs/devbie/acceptance/AcceptanceTest.java:
```diff
@@ -45,8 +46,10 @@ void setUp() throws JsonProcessingException {
         RestAssured.port = port;
 
         userId = createUser();
+        //Todo: Role.USER 테스트도 추가하기
```
Develop 브랜치에 Todo가 올라가는건 안좋다고 생각합니다.

c-yeonwoo reviewed (approved):
코드보다가 생각났는데 다음 추가 이슈로, User에 role이라는 필드가 추가된 만큼, user를 사용하고 있는 question, answer, recommendation에서도
테스트할 때 수정이 필요하지 않을까 싶네요ㅎㅎ 
수고하셨습니다!

KimSeongGyu1 commented on src/main/java/underdogs/devbie/notice/controller/NoticeController.java:
```diff
@@ -29,7 +31,10 @@
 
     @PostMapping
-    public ResponseEntity<Void> save(@Valid @RequestBody NoticeCreateRequest request) {
+    public ResponseEntity<Void> save(
+        @AdminUser Boolean isAdmin,
+        @Valid @RequestBody NoticeCreateRequest request
+    ) {
```
Resolver로 isAdmin만 넘겨주고 사용하지 않는다면 Resolver를 사용할 이유가 있을까요?

interceptor의 path를 잘 걸어주고 interceptor만 이용하는 것이 더 적절하지 않을까요?

Resolver로 넘겨준다면 더 유의미한 값을 넘겨주고 넘겨준 값을 활용해야 한다고 생각합니다

sonypark replied:
동의합니다. 저도 `@LoginUser`와 비슷한 방식으로 `@AdminUser`를 활용할 수 있지 않을까 해서 위와 같이 구현했는데, 사실상 현재 NoticeController에서 user 관련 로직이 필요가 없습니다. 그래서 Interceptor 만으로 권한 확인은 충분할 것 같네요.

KimSeongGyu1 replied:
더 생각을 해봤는데 User 객체에 isAdmin() 메서드를 추가하면
@LoginUser를 사용할 수 있을 것 같아요

Admin에 관련된 interceptor를 따로 생성할 것인지, Controller에서 @LoginUser로 받아서 isAdmin()으로 검사할 것인지 고민해볼 만 한 것 같아요

KimSeongGyu1 reviewed (commented):
고생하셨어요

고민해볼만한 거리가 있는 것 같아서 한가지 코멘트 남겼어요

sonypark committed:
refactor: @AdminUser 리졸버는 불필요하다고 판단하여 삭제

- 현재는 interceptor에서 검증하는 것만으로 충분하다고 판단

sonypark committed:
refactor: 로그인 관련 컨트롤러 인증 로직을 @novalidate를 사용하는 방식으로 변경

hongbin-dev reviewed (commented):
소니~ 수고하셨어요
간단한 리뷰 남겨용

sonypark committed:
refactor: Admin 대신 Role 어노테이션 추가

sonypark committed:
chore: TODO 주석 삭제

sonypark committed:
refactor: 권한이 없을 경우 예외 발생

- 기존에 인터셉터에서 권한 검증해서 권한이 없으면 false를 리턴하는 방식으로 구현했다.
- 이때 문제는 예외가 아니라 200이 리턴되었다.
- 따라서 권한이 없을 경우 인터셉터에서 예외를 발생하도록 변경했다.

KimSeongGyu1 commented on src/main/java/underdogs/devbie/auth/controller/resolver/LoginUserArgumentResolver.java:
```diff
@@ -30,7 +30,6 @@ public User resolveArgument(MethodParameter parameter, ModelAndViewContainer mav
-        // todo: 구지 있어야할까?
         if (StringUtils.isEmpty(userId)) {
             throw new IllegalArgumentException();
         }
```
빈 값에 대해서 적절한 에러가 발생하는 것을 테스트에서 확인했으니
주석만 지우지 말고 if 문 통채로 지워도 될 것 같습니다

sonypark replied:
네 알겠습니다~

KimSeongGyu1 commented on src/main/java/underdogs/devbie/config/AuthConfig.java:
```diff
@@ -20,9 +20,7 @@
-            .addPathPatterns("/api/**")
-            .excludePathPatterns("/api/auth/login", "/api/auth/login-url")
-            .excludePathPatterns("/api/notices", "/api/notices/**");
+            .addPathPatterns("/api/**");
```
excludePath 대신 어노테이션을 사용하는 것이
저희 팀 전체적으로 합의가 된 내용인가요?
저는 다같이 합의 했던 기억이 없어요

sonypark replied:
아뇨 다같이 합의하지는 않았습니다. 
보스독,앨런과 이야기를 나눈 후 excludePath보단 어노테이션을 사용하는 방식이 더 편이성이 높다고 생각해서 수정해보았습니다.
리뷰 과정에서 다른 의견이 나오면 다시 논의할 예정이었습니다.
유안은 excludePath와 어노테이션을 사용하는 방식 중 어느 쪽이 더 낫다고 생각하시나요?

KimSeongGyu1 replied:
저도 어노테이션 쪽이 더 편하다고 생각합니다
하지만 다 같이 신경써야 하는 부분에 팀적이 합의가 사전에 이루어 지지 않으면 혼동의 여지가 있을 것 같아 말씀드렸습니다

KimSeongGyu1 commented on src/main/java/underdogs/devbie/auth/controller/interceptor/BearerAuthInterceptor.java:
```diff
@@ -25,8 +26,15 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
+        String userId = claims.get("userId").toString();
+        String role = claims.get("role").toString();
+
+        if (interceptorValidator.hasRoleAnnotation(handler)) {
+            interceptorValidator.validateRole(handler, role);
+        }
```
```suggestion
        if (interceptorValidator.isValidRole(handler, role)) {
            return true;
        }
```
이런식으로 변경하면 위에서 사용한 if 문과 통일성 있게 더 깔끔해지지 않을까요?

sonypark replied:
그렇네요. 말씀하신대로 수정했습니다! 👍

KimSeongGyu1 commented on src/main/java/underdogs/devbie/auth/controller/interceptor/utils/InterceptorValidator.java:
```diff
@@ -1,15 +1,34 @@
+        return Objects.nonNull(methodAnnotation);
+    }
+
+    public void validateRole(Object handler, String role) {
+        Role methodAnnotation = ((HandlerMethod)handler).getMethodAnnotation(Role.class);
+        RoleType roleType = RoleType.valueOf(role);
+
+        if (!Arrays.asList(methodAnnotation.role())
+            .contains(roleType)) {
+            throw new InvalidAuthorizationException();
+        }
+    }
```
Role의 값들을 꺼내서 검사하는 것 보다는
Role 클래스에 validate 책임을 맡기는 것이 더 객체지향적이지 않을까요?

sonypark replied:
@Role 인터페이스에서는 디폴트 메서드에 인자를 받을 수 없어 위와 같이 구현했습니다. :)

KimSeongGyu1 commented on src/main/java/underdogs/devbie/notice/controller/NoticeController.java:
```diff
@@ -47,20 +54,25 @@
+    @Role(role = {RoleType.ADMIN})
     @DeleteMapping("/{id}")
-    public ResponseEntity<Void> delete(@PathVariable Long id) {
+    public ResponseEntity<Void> delete(
+        @PathVariable Long id
+    ) {
```
파라미터가 하나일 때 이 부분은 여러 라인을 사용했지만
77번째 줄은 한 줄을 사용했네요

통일성이 있으면 좋을 것 같아요
저는 개인적으로 파라미터가 한개면 한줄에 써도 될 것 같아요

다른 곳에도 불규칙하게 쓰인 곳이 조금 보이네요

sonypark replied:
isAdmin 어노테이션을 안 쓰는 방향으로 수정하는 과정에서 포맷을 맞추지 않았네요.. 수정했습니다!

KimSeongGyu1 reviewed (commented):
새로운 내용을 적용하느라 고생하셨어요

의견 남겼으니 확인 부탁드려요

sonypark committed:
refactor: Interceptor에서 권한 인증하는 로직 수정

sonypark committed:
refactor: 불필요한 로직 삭제

slowCoyle commented on src/main/java/underdogs/devbie/auth/dto/UserTokenDto.java:
```diff
@@ -4,16 +4,19 @@
+
     public static UserTokenDto from(User user) {
-        return new UserTokenDto(user.getId());
+        return new UserTokenDto(user.getId(), user.getRoleType());
```
Dto에서는 new로 생성하는 것을 허용하기로 했나요 ? 
컨벤션이 확실히 기억안나네용

그리고 두 개 이상의 매개변수일 경우에는 user를 매개변수로 던진 뒤 Dto에서 get하는게 낫지 않을까용?

sonypark replied:
DTO의 정적메서드에서 자기 자신을 new로 생성하는건 필요하지 않을까요??
다만 생성자가 하나일경우 new 생성자를 쓰기로 했고 두 개 이상일 경우는 builder 패턴으로 하기로 했는데, 위 코드에서는 지키지 않았네요.
아마 저 부분은 첫 주에 다같이 짠 코드인데, 위 원칙을 세운게 그 이후라서 반영이 안 된 것 같아요!
그래서 builder 패턴으로 수정했습니다.

```java
    public static UserTokenDto from(User user) {
        return UserTokenDto.builder()
            .id(user.getId())
            .roleType(user.getRoleType())
            .build();
    }
```

slowCoyle commented on src/main/java/underdogs/devbie/notice/controller/NoticeController.java:
```diff
@@ -28,6 +31,7 @@
 
     private final NoticeService noticeService;
 
+    @Role(role = {RoleType.ADMIN})
```
위처럼 어노테이션에 role이라는 값을 주고 싶을 때는 어떻게 하신건가요 ? 

Role 어노테이션 인터페이스에 role() 메서드가 존재하면 위와 같이 작성할 수 있는 건가요 ?

sonypark replied:
네 Role 인터페이스에 다음과 같이 디폴트 메서드를 주는 방식으로 구현했습니다.

```java
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface Role {

    RoleType[] role() default {RoleType.ADMIN, RoleType.USER, RoleType.GUEST};
}
```

sonypark commented on src/main/java/underdogs/devbie/auth/controller/interceptor/utils/InterceptorValidator.java:
```diff
@@ -1,15 +1,31 @@
+        Role methodAnnotation = ((HandlerMethod)handler).getMethodAnnotation(Role.class);
+        RoleType roleType = RoleType.valueOf(role);
+
+        if (!Arrays.asList(methodAnnotation.role())
+            .contains(roleType)) {
```
```java
@Target({ElementType.METHOD})
@Retention(RetentionPolicy.RUNTIME)
public @interface Role {

    RoleType[] role() default {RoleType.ADMIN, RoleType.USER, RoleType.GUEST};
}
```

위 코드에서 디폴트 메서드에 인자를 받을 수 없더라구요.. 그래서 위와 같이 구현했습니다. 혹시 isNotExistRoleType과 같은 메서드를 만들어 로직을 수행할 수 있는 방법을 아신다면 알려주시면 감사하겠습니다!

sonypark committed:
refactor: UserTokenDto 정적메서드의 생성 패턴을 빌더 패턴으로 변경

sonypark committed:
conflict: 충돌 해결

sonypark committed:
Merge branch 'develop' of https://github.com/woowacourse-teams/2020-devbie into feature/85-add-role-for-authorization

sonypark committed:
fix: @Role 어노테이션이 없는 컨트롤러의 경우 처리 로직 추가

- @Role 어노테이션이 없는 컨트롤러의 경우 처리 로직이 없어서 405 에러 발생
